cirq/circuits/expectation_value.py

```diff
@@ -0,0 +1,340 @@
+                 in calculating the expectation value of operator
+
+        operator: Operator input as a dictionary where keys are
+                    cirq.ops.PauliString type and values are coefficients
+                  example:
+                  paulistring = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.X, qubits[1]: (cirq.Pauli.Z)})
+                  paulistring2 = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.Z, qubits[2]: (cirq.Pauli.Y)})
+                  operator = {paulistring: 1.234, paulistring2: -5.678}
+
+                sampling is for future incorporation with QPU (not simulator)
+
+        n_samples: number of runs for sampling
+
+        quadratic_z: Boolean that determines whether operator only has terms
+                     quadratic in Pauli Z
+                     If true, computation is faster
+
+    Returns:
+        expectation value of operator argument
+
+    """
+
+    qubits = list(circuit.all_qubits())
+    sim = XmonSimulator()
+
+    # remove measurements from circuit such that appropriate rotations
+    # can be applied, measurements are added later
+    circuit = no_meas(circuit)
+
+    # For general operators, the expectation value by sampling
+    # over n_samples circuit runs is computed here
+    if not quadratic_z:
+
+        circuit = no_meas(circuit)
+
+        expectation = 0
+        for term, coef in operator.items():
+
+            # check if identity:
+            identity_coeficient = 0.
+            if term == ():
+                identity_coeficient = coef
+                continue
+
+            # Add appropriate rotation gates to circuit
+            # and add measurement gates
+            # dictionary of qubits to be measured and its indices
+            q_dict = dict()
+
+            for qubit, pauli in term.items():
+
+                if pauli == Pauli.X:
+                    circuit.append(RotYGate(half_turns=-1 / 2).on(
+                        qubit))
+
+                elif pauli == Pauli.Y:
+                    circuit.append(RotXGate(half_turns=+1 / 2).on(
+                        qubit))
+
+                # After rotation we can re-add measurement gates
+                q_dict[qubit] = qubits.index(qubit)
+
+            # Add Measurement gates:
+            circuit = measurement_gates(circuit, q_dict)
+
+            # run circuit:
+            results = sim.run(circuit, repetitions=n_samples)
+
+            operator_meas = 1
+
+            for qubit_key in results.measurements.keys():
+                operator_meas *= results.measurements[qubit_key][:, 0].astype(
+                    int) * (-2) + 1
+
+            mean_measurement = np.mean(operator_meas)
+
+            expectation += mean_measurement * coef
+
+        # add identity term:
+        expectation += identity_coeficient
+
+        return expectation
+
+    if quadratic_z:
+
+        # Check whether circuit already has measurement gates:
```

Comment:
Extract this giant if-block into a method of its own (`_expectation_from_sampling_assuming_quadratic`?).